src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -3068,6 +3073,155 @@ impl Coordinator {
         Ok(ExecuteResponse::AlteredObject(ObjectType::Secret))
     }
 
+    #[instrument]
+    pub(super) async fn sequence_alter_sink_prepare(
+        &mut self,
+        ctx: ExecuteContext,
+        plan: plan::AlterSinkPlan,
+    ) {
+        // 1. Put a read hold on the new relation
+        let id_bundle = crate::CollectionIdBundle {
+            storage_ids: BTreeSet::from_iter([plan.sink.from]),
+            compute_ids: BTreeMap::new(),
+        };
+        let read_hold = self.acquire_read_holds(&id_bundle);
+
+        let Some(read_ts) = read_hold.least_valid_read().into_option() else {
+            ctx.retire(Err(AdapterError::UnreadableSinkCollection));
+            return;
+        };
+
+        let otel_ctx = OpenTelemetryContext::obtain();
+
+        let plan_validity = PlanValidity {
+            transient_revision: self.catalog().transient_revision(),
+            dependency_ids: BTreeSet::from_iter([plan.sink.from]),
+            cluster_id: Some(plan.in_cluster),
+            replica_id: None,
+            role_metadata: ctx.session().role_metadata().clone(),
+        };
+
+        // Re-resolve items in the altered statement
+        // Parse statement.
+        let create_sink_stmt = match mz_sql::parse::parse(&plan.sink.create_sql)
+            .expect("invalid create sink sql")
+            .into_element()
+            .ast
+        {
+            Statement::CreateSink(stmt) => stmt,
+            _ => unreachable!("invalid statment kind for sink"),
+        };
+        let catalog = self.catalog().for_system_session();
+        let (_, resolved_ids) = match mz_sql::names::resolve(&catalog, create_sink_stmt) {
+            Ok(ok) => ok,
+            Err(e) => {
+                ctx.retire(Err(AdapterError::internal("ALTER SINK", e)));
+                return;
+            }
+        };
+
+        // Now we must wait for the sink to make enough progress such that there is overlap between
+        // the new `from` collection's read hold and the sink's write frontier.
+        self.controller.install_storage_watch_set(
+            BTreeSet::from_iter([plan.id]),
+            read_ts,
+            Box::new(WatchSetResponse::AlterSinkReady(AlterSinkReadyContext {
+                ctx,
+                otel_ctx,
+                plan,
+                plan_validity,
+                resolved_ids,
+                read_hold,
+            })),
+        );
+    }
+
+    #[instrument]
+    pub async fn sequence_alter_sink_finish(
+        &mut self,
+        session: &mut Session,
+        plan: plan::AlterSinkPlan,
+        resolved_ids: ResolvedIds,
+        read_hold: ReadHolds<Timestamp>,
+    ) -> Result<ExecuteResponse, AdapterError> {
+        let plan::AlterSinkPlan {
+            id,
+            sink,
+            with_snapshot,
+            in_cluster,
+        } = plan;
+
+        // Assert that we can recover the updates that happened the timestamps of the write
```

Comment:
```suggestion
        // Assert that we can recover the updates that happened before the timestamps of the write
```

Reply:
The comment is indeed wrong but the correct comment wants to say "at the timestamps of the write frontier". Fixed!